### ISSUE TITLE
Change cache directory structure?

### ISSUE THREAD

**Issue by jrosdahl** [ccache/ccache#416]
The idea of changing the directory structure comes up from time to time. Let's discuss it separately in this issue.

References:

- #168
- Discussion in #218:
  - https://github.com/ccache/ccache/issues/218#issuecomment-487701484
  - https://github.com/ccache/ccache/issues/218#issuecomment-487734028
- Many comments in #404

**Comment by pavsa** [comment]
There's definitely a problem with lock waiting on stats file when doing parallel builds with more cores. I am not sure by how much since you only see it with log enabled, which itself is slows down the build. 

That also reminds me, if you enable logging to file, it is unreadable if you have 20 parallel instances writing to the same log. I had to write a custom script to replay it back so that single ccache instances are in the log serially. 
Maybe output caching and print it all at the end with the lock on the log file?

**Comment by afbjorklund** [comment]
You could try the tracing mode instead, which was intended for measuring things such as this one...
I think in the future that stats should be separated from cache, and that we need to rethink the locks.

The log file entries should all be separated by PID, but suggest the debug mode if you need "readable".
Most likely using syslog is faster, but I haven't really benchmarked - just trusting what other people say.

**Comment by jrosdahl** [comment]
Changing the directory structure in a backward incompatible way is a bit problematic. With backward compatibility I'm referring to the scenario where old ccache versions run against the same cache directory as newer versions do: old versions should not interfere with stuff put in the cache by new versions, and vice versa. Specifically, cleanup performed by old versions should work with stuff put there by newer versions.

Examples of things that will break backward compatibility:

* Changing names of the first sublevel directories to something else than `[0-9a-f]`.
* Moving _cache size_ and _files in cache_ information to another place than in `$CCACHE_DIR/[0-9a-f]/stats`.

If maintaining backward compatibility is too hard or is too restraining, it's of course possible to solve the problem in a more abrupt way, for instance by letting new versions refuse to work against an old-style cache directory. But I would like to avoid that if possible, and I think that we can come a long way while maintaining backward compatibility, as indicated in #417.

From https://github.com/ccache/ccache/issues/404#issuecomment-489415533:

> @jrosdahl mentions that he has some secret special plans for improving cleanup

Well, not exactly secret plans in the sense "intentionally not telling", only in the sense "so far only exist in my head"...

> so maybe stats too ?

Yes, my improvement idea will reduce lock contention for cache hits but not for cache misses.

**Comment by afbjorklund** [comment]
There are no details added to #414 yet, so I am not sure how the storage backends are going to work...

But my plan was to add the file-based implementation to a new top-level directory, separate from the old.
In this new directory, there could be a couple of database files or there could be multiple subdirectories.

At any rate, I think updating the stats is going to have to use a separate API from the storage API ?

---

So I think there will be a mild cache upgrade (with #411), that is still compatible with the current setup...

For the storage backends (in #414) it won't be, as it might not actually use the disk for storage at all.

**Comment by pavsa** [comment]
If the new system is better, I think you should break the compatibility and not worry about it. There's a *lot* of work involved in maintaining backward compatibility which could be better spent on new features/fixing bugs. 

You can release it as a new major version with notice that it is not backwards compatible. It is not even now, the manifests are now unreadable by older versions.

**Comment by jrosdahl** [comment]
> At any rate, I think updating the stats is going to have to use a separate API from the storage API ?

I haven't thought much about changes to the stats system other than that counter updates shouldn't be part of the storage API. It the current stats implementation is perceived as a problem, I'm all for making alternative "stats backends" as well.

**Comment by jrosdahl** [comment]
@afbjorklund wrote:
> But my plan was to add the file-based implementation to a new top-level directory, separate from the old.

@pavsa wrote:
> If the new system is better, I think you should break the compatibility and not worry about it. [...] You can release it as a new major version with notice that it is not backwards compatible.

As I see it, that's not an option at all. If compatibility is broken, it's a new tool that needs to have another name, use other environment variables and default directory locations, etc. Or: Keep the old name but make new versions bail out as soon as they see a cache directory created by an old version, thus forcing you to make sure that old versions never touch a directory created by the new version (painful). Creating a new tool (ccache-ng, ccache2 or whatever) would of course be fine, but it's not something I'm interested in.

I do understand why one might think that breaking compatibility is not a big issue: when you have full control over your environment, it's not a big deal. But there are other scenarios, for instance this: Say that you have a CI system that exposes a cache directory to builds of different softwares and/or branches of the same software. The cache directory is shared because you want a size limit per build machine, so it doesn't make sense to keep one cache directory for each software or software branch. Then the actual ccache version to use is stored with the software to build (it could be part of a build container image associated with the software). If I can avoid it, I don't want to break this kind of usage that has worked fine for many years. (A side note: I've already broken it once in ccache 3.2 by moving the max size/files configuration to a config file. This has unfortunate consequences for Android builds that share a cache with other software builds since Android still happens to use (or at least quite recently used to use) a prebuilt ccache 3.1.9.)

This is my current view (not saying it won't change in the future).

Does this make sense to you?

> It is not even now, the manifests are now unreadable by older versions.

That's not at all the same issue. Old versions are not bothered by unreadable manifest or results from newer versions since they will never want to read them, only clean them up.

**Comment by pavsa** [comment]
> 
> 
> @afbjorklund wrote:
> 
> > But my plan was to add the file-based implementation to a new top-level directory, separate from the old.
> 
> @pavsa wrote:
> 
> > If the new system is better, I think you should break the compatibility and not worry about it. [...] You can release it as a new major version with notice that it is not backwards compatible.
> 
> As I see it, that's not an option at all. If compatibility is broken, it's a new tool that needs to have another name, use other environment variables and default directory locations, etc. Or: Keep the old name but make new versions bail out as soon as they see a cache directory created by an old version, thus forcing you to make sure that old versions never touch a directory created by the new version (painful). Creating a new tool (ccache-ng, ccache2 or whatever) would of course be fine, but it's not something I'm interested in.
> 
> I do understand why one might think that breaking compatibility is not a big issue: when you have full control over your environment, it's not a big deal. But there are other scenarios, for instance this: Say that you have a CI system that exposes a cache directory to builds of different softwares and/or branches of the same software. The cache directory is shared because you want a size limit per build machine, so it doesn't make sense to keep one cache directory for each software or software branch. Then the actual ccache version to use is stored with the software to build (it could be part of a build container image associated with the software). If I can avoid it, I don't want to break this kind of usage that has worked fine for many years. (A side note: I've already broken it once in ccache 3.2 by moving the max size/files configuration to a config file. This has unfortunate consequences for Android builds that share a cache with other software builds since Android still happens to use (or at least quite recently used to use) a prebuilt ccache 3.1.9.)
> 
> This is my current view (not saying it won't change in the future).
> 
> Makes sense?
> 
> > It is not even now, the manifests are now unreadable by older versions.
> 
> That's not at all the same issue. Old versions are not bothered by unreadable manifests or results from newer versions since they will never want to read them, only clean them up.

I see. Then I think the best solution is probably the one you proposed - new directory inside ccache_dir that won't be bothered by old versions.

**Comment by jrosdahl** [comment]
> I see. Then I think the best solution is probably the one you proposed - new directory inside ccache_dir that won't be bothered by old versions.

I haven't proposed that (why did you think I did?). Such a solution has the problem that the max cache size as configured in ccache.conf will be exceeded since new versions will create up to the max size of data in the new directory and similar for old versions in the old directory structure.

**Comment by afbjorklund** [comment]
My comment was mostly in the light of #414, once we "give up" the storage of the cache to a pluggable system there is no dictating what the layout of that storage will be... So it **will** be external to the current cache, which could still hold the config (or even the stats, since that most likely will be a separate topic).

For the current cache directory structure (`0-f`), then I think #408 is the closest we have. It doesn't really change anything for the directory structure, so should be drop-in replaceable - even though the contents won't be shared. But if you run a clean with old ccache on the new cache dir, it will remove the `.result`

One could write a simple tool to convert/upgrade the cache format, but I am not sure if it will be needed.

There are other pending changes (e.g. compression) that are likely to invalidate the contents anyway...

----

But you are right, if keep the old files and the new files in the same directory they **will** use more space.

As in if you set them both to use 5G, then that would probably make the total directory size 10G instead.
(or even more, if you decide to use more than one storage plugin and these aren't aware of each other)

Not sure it is a big problem though, disks are big these days (and _one_ build is like 20G uncompressed...)

**Comment by afbjorklund** [comment]
Here is what `$CACHE_DIR` looks like, when using `$CCACHE_MEMCACHED_ONLY`:

```
├── 0
│   └── stats
├── 1
│   └── stats
├── 2
│   └── stats
├── 3
│   └── stats
├── 4
│   └── stats
├── 5
│   └── stats
├── 6
│   └── stats
├── 7
│   └── stats
├── 8
│   └── stats
├── 9
│   ├── a
│   └── stats
├── a
│   └── stats
├── b
│   └── stats
├── c
│   └── stats
├── ccache.conf
├── d
│   └── stats
├── e
│   └── stats
├── f
│   ├── 6
│   └── stats
└── tmp
```

The actual contents are in the database:

```
$ memcdump 
f644fc05e414242e798c502dcefdfdfc-130
9a324419c4dc84562337280f1bc94e3e-215
```

Those created second-level directories are probably a bug, but that's another story...
Also I think the manifest files _require_ files at the moment, but that could be fixed too.

----

My plan was to keep the new storage inside a _hidden_ directory in the cache directory.
The contents of that would be internal to the implementation, including subdirectories.

Depending on the storage plugin, it could be just a few files (like for a local database).
Or it could be a lot of small files and subdirectories, similar to what we currently have.

**Comment by pavsa** [comment]
> 
> 
> > I see. Then I think the best solution is probably the one you proposed - new directory inside ccache_dir that won't be bothered by old versions.
> 
> I haven't proposed that (why did you think I did?). 

Oh, sorry, I misread the quote, it was @afbjorklund :) :

@afbjorklund wrote:

    But my plan was to add the file-based implementation to a new top-level directory, separate from the old.

> Such a solution has the problem that the max cache size as configured in ccache.conf will be exceeded since new versions will create up to the max size of data in the new directory and similar for old versions in the old directory structure.

One solution would be to fail if the old ccache structure is detected and to add one new variable to the new ccache version (for example lets call the old dir structure V1: ```CCACHE_IGNORE_DIR_STRUCTURE_V1```) that would stop it from failing.

It would be very clearly stated in the docs that if this variable is enabled, statistics for the old structure would not be updated or monitored and it will result in doubling the size of the cache.

**Comment by jrosdahl** [comment]
@afbjorklund wrote:
> My comment was mostly in the light of #414, once we "give up" the storage of the cache to a pluggable system there is no dictating what the layout of that storage will be... So it **will** be external to the current cache, which could still hold the config (or even the stats, since that most likely will be a separate topic).

Well, even if the current file system backend will be expressed in terms of a plugin API, there is still the question of which backend will be the **default**. And my current view is that it should be the current "naive file system" backend, but improved in a reasonably backward compatible way.

> [...] But if you run a clean with old ccache on the new cache dir, it will remove the `.result`

Exactly.

> One could write a simple tool to convert/upgrade the cache format, but I am not sure if it will be needed.

Definitely not needed.

> Not sure it is a big problem though, disks are big these days (and _one_ build is like 20G uncompressed...)

At work, we use up to 5% (depending on build node) of storage per node for ccache. Doubling that space would force us to have to partition stuff in another way or come up with other workarounds. It's not as easy as "disks are big and cheap".

**Comment by afbjorklund** [comment]
If compatibility is required, then the .result files are probably the way to go. Even though the "doubling" should only happen if you use multiple versions of ccache (old/new) towards the same cache directory...

**Comment by jrosdahl** [comment]
> If compatibility is required, then the .result files are probably the way to go. Even though the "doubling" should only happen if you use multiple versions of ccache (old/new) towards the same cache directory...

I wouldn't say 100% required, it's just my spider-sense telling me that it's probably a good idea to keep compatibility if it isn't too painful to do so.
 
> Even though the "doubling" should only happen if you use multiple versions of ccache (old/new) towards the same cache directory...

Yes, that's essentially what's our CI system does. And I suspect that we are not the only ones; [there is always a stranger use case](https://xkcd.com/1172/).

**Comment by jrosdahl** [comment]
Closing this issue now since the discussion seems to have ended.

**Comment by pavsa** [comment]
What's the result? What will be implemented?

**Comment by jrosdahl** [comment]
> What's the result? What will be implemented?

I don't think that we should make structural changes of the default (currently only) backend storage layout, with the exception of improving how statistics values are updated/stored (related to #417, description still TBD unfortunately). Alternative backends could do something completely different. Maybe an alternative backend could become the default one in the future when it has matured.
